src/herder/TxSetFrame.cpp

```diff
@@ -1549,17 +1562,296 @@ TxSetPhaseFrame::sortedForApply(Hash const& txSetHash) const
+{
+    auto const& lcl = app.getLedgerManager().getLastClosedLedgerHeader();
+    // Verify the fee map for the phase. This check is independent of the phase
+    // type or contents.
+    if (!checkFeeMap(getInclusionFeeMap(), lcl.header))
+    {
+        return false;
+    }
+
+    bool isSoroban = mPhase == TxSetPhase::SOROBAN;
+
+    // Ensure that the phase contains only the transactions of expected
+    // kind (Soroban or classic).
+    for (auto const& tx : *this)
+    {
+        if (tx->isSoroban() != isSoroban)
+        {
+            CLOG_DEBUG(Herder,
+                       "Got bad generalized txSet with invalid "
+                       "phase {} transactions",
+                       static_cast<size_t>(mPhase));
+            return false;
+        }
+    }
+
+    // Then check the phase-specific properties. This may rely on transactions
+    // belonging to the valid phase.
+    bool checkPhaseSpecific =
+        isSoroban
+            ? checkValidSoroban(
+                  lcl.header,
+                  app.getLedgerManager().getSorobanNetworkConfigReadOnly())
+            : checkValidClassic(lcl.header);
+    if (!checkPhaseSpecific)
+    {
+        return false;
+    }
+
+    return txsAreValid(app, lowerBoundCloseTimeOffset,
+                       upperBoundCloseTimeOffset);
+}
+
+bool
+TxSetPhaseFrame::checkValidClassic(LedgerHeader const& lclHeader) const
+{
+    if (isParallel())
+    {
+        CLOG_DEBUG(Herder, "Got bad txSet: classic phase can't be parallel");
+        return false;
+    }
+    if (this->size(lclHeader) > lclHeader.maxTxSetSize)
+    {
+        CLOG_DEBUG(Herder, "Got bad txSet: too many classic txs {} > {}",
+                   this->size(lclHeader), lclHeader.maxTxSetSize);
+        return false;
+    }
+    return true;
+}
+
+bool
+TxSetPhaseFrame::checkValidSoroban(
+    LedgerHeader const& lclHeader,
+    SorobanNetworkConfig const& sorobanConfig) const
+{
+    bool needParallelSorobanPhase = protocolVersionStartsFrom(
+        lclHeader.ledgerVersion, PARALLEL_SOROBAN_PHASE_PROTOCOL_VERSION);
+    if (isParallel() != needParallelSorobanPhase)
+    {
+        CLOG_DEBUG(Herder,
+                   "Got bad txSet: Soroban phase parallel support "
+                   "does not match the current protocol; '{}' was "
+                   "expected",
+                   needParallelSorobanPhase);
+        return false;
+    }
+    // Ensure the total resources are not over ledger limit.
+    auto totalResources = getTotalResources();
+    if (!totalResources)
+    {
+        CLOG_DEBUG(Herder, "Got bad txSet: total Soroban resources overflow");
+        return false;
+    }
+
+    auto maxResources = sorobanConfig.maxLedgerResources();
+
+    if (protocolVersionStartsFrom(lclHeader.ledgerVersion,
+                                  PARALLEL_SOROBAN_PHASE_PROTOCOL_VERSION))
+    {
+        maxResources.setVal(Resource::Type::INSTRUCTIONS,
```

Comment:
Nit: Maybe add a comment explaining that we'll validate instructions later in this function?

Reply:
Good catch, added.